### ISSUE TITLE
Rule  S4035: False positive for abstract declarations

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#2322]
### Description

```csharp
public abstract class ValueObject<T> : IEquatable<T> where T : ValueObject<T>
{
    public abstract bool Equals(T other); // compliant
}

public class MyValueObject : ValueObject<MyValueObject>
{
    public override bool Equals(MyValueObject other)
    {
        return RefrenceEquals(this, other); //non-compliant
    }
}

public sealed class MySealedValueObject : ValueObject<MySealedValueObject >
{
    public override bool Equals(MySealedValueObject  other)
    {
        return RefrenceEquals(this, other); //compliant
    }
}
```

To me, this should not raise the this concern on the abstract class with an abstract method. The implementing class should, if not sealed.

### Related information
* #461
* 7.11.0.8083
* Visual Studio 15.9.7

**Comment by christophe-zurn-sonarsource** [comment]
Hello @Corniel,
Indeed, for the same reason Rule S4035 does not raise an issue for `virtual` `Equals` method, it should also not raise issue for `abstract` `Equals` method.

```
    public abstract class ValueObject<T> : IEquatable<T> where T : ValueObject<T>  // FP
    {
        public abstract bool Equals(T other);
    }
```
